### ISSUE TITLE
error in compareSteady

### ISSUE THREAD

**Issue by slebedeva** [ste-depo/INSPEcT#28]
Hi! I have encountered an error when trying to run function compareSteady:

```
> diffrates0_48 <- compareSteady(SteadyObj[,c("0","48")])
Comparative analysis of the 72679 with both intronic and exonic signals.
Evaluating model no-reg [1/8]...
Evaluating model s [2/8]...
Evaluating model p [3/8]...
Evaluating model d [4/8]...
Evaluating model sp [5/8]...
Evaluating model sd [6/8]...
Evaluating model pd [7/8]...
Evaluating model spd [8/8]...
Comparative analysis of the 19559 without intronic signals.
Evaluating model no-reg [1/4]...
Evaluating model s [2/4]...
Evaluating model d [3/4]...
Evaluating model sd [4/4]...
Error in diz[apply(AIC, 1, which.min)] : invalid subscript type 'list'
```

I am on Linux with R version 3.6.2 and INSPEcT" v. "1.16.3".
Thanks a lot for looking into this!

**Comment by ste-depo** [comment]
Hi, can you please help me with the debug?
You should rerun the command after typing on the terminal the following command:
> options(error=browser)
After the error occurs, the interactive debugging starts. Now type:
(browser>) class(AIC)
and
(browser>) str(AIC)
and return back to me the results. Then go back to the normal error handling typing this command:
> options(error=NULL)

You can also do this on a subset of the dataset, i.e. compareSteady(SteadyObj[1:1000,c("0","48")] ,even though subsetting could eliminate the gene - or genes - giving problems, in that case try a new subset.

**Comment by slebedeva** [comment]
Hi Stefano, thanks a lot for superfast reply!!
Here the result:

```
Error in diz[apply(AIC, 1, which.min)] : invalid subscript type 'list'
Called from: compareSteady(SteadyObj[, c("0", "48")])
Browse[1]> class(AIC)
[1] "standardGeneric"
attr(,"package")
[1] "methods"
Browse[1]> str(AIC)
Formal class 'standardGeneric' [package "methods"] with 8 slots
  ..@ .Data     :function (object, ..., k = 2)  
  ..@ generic   : chr "AIC"
  .. ..- attr(*, "package")= chr "stats"
  ..@ package   : chr "stats"
  ..@ group     : list()
  ..@ valueClass: chr(0) 
  ..@ signature : chr [1:2] "object" "k"
  ..@ default   :Formal class 'derivedDefaultMethod' [package "methods"] with 4 slots
  .. .. ..@ .Data  :function (object, ..., k = 2)  
  .. .. ..@ target :Formal class 'signature' [package "methods"] with 3 slots
  .. .. .. .. ..@ .Data  : chr "ANY"
  .. .. .. .. ..@ names  : chr "object"
  .. .. .. .. ..@ package: chr "methods"
  .. .. ..@ defined:Formal class 'signature' [package "methods"] with 3 slots
  .. .. .. .. ..@ .Data  : chr "ANY"
  .. .. .. .. ..@ names  : chr "object"
  .. .. .. .. ..@ package: chr "methods"
  .. .. ..@ generic: chr "AIC"
  .. .. .. ..- attr(*, "package")= chr "stats"
  ..@ skeleton  : language (new("derivedDefaultMethod", .Data = function (object, ..., k = 2)  UseMethod("AIC"), target = new("signature", .| __truncated__ ...
```

**Comment by ste-depo** [comment]
Hi again, 

sorry but that attempt didn't work. There should be some gene (or genes) causing the problem. Can you identify a small subset of the dataset that generate the error and attach it to this conversation?

Sorry for the late reply, this time..

Stefano

**Comment by slebedeva** [comment]
Hi Stefano! Unfortunately, when I try on a subset of genes, a different error pops up, so it is hard to say if these genes contain the problem. 

```
compareSteady(testObj[,c("0","48")])
Comparative analysis of the 5 with both intronic and exonic signals.
Evaluating model no-reg [1/8]...
Evaluating model s [2/8]...
Evaluating model p [3/8]...
Evaluating model d [4/8]...
Evaluating model sp [5/8]...
Evaluating model sd [6/8]...
Evaluating model pd [7/8]...
Evaluating model spd [8/8]...
Comparative analysis of the 1 without intronic signals.
Evaluating model no-reg [1/4]...
Error in apply(outK_K[, 1:2] < 0, 1, any) : 
  dim(X) must have a positive length
In addition: Warning messages:
1: In min(wt_rates[wt_rates[, 1] != 0 & !is.na(wt_rates[, 1]), 1]) :
  no non-missing arguments to min; returning Inf

## this gives no error:
compareSteady(testObj[,c("0","12")])
```

I attach the test object.
[SteadyObj_forStefano.zip](https://github.com/ste-depo/INSPEcT/files/5006828/SteadyObj_forStefano.zip)


Thank you again!
Svetlana

**Comment by ste-depo** [comment]
Hi Svetlana,

I think I have resolved the issue! The problem was due to 2 genes that could not be resolved properly by the modeling. The code now checks for this kind of issue and eventually remove the genes causing the problem. Here is the output of the code on the dataset that you provided to me:

> diffrates0_48 <- compareSteady(SteadyObj[,c("0","48")])
Comparative analysis of the 72679 with both intronic and exonic signals.
Evaluating model no-reg [1/8]...
Evaluating model s [2/8]...
Evaluating model p [3/8]...
Evaluating model d [4/8]...
Evaluating model sp [5/8]...
Evaluating model sd [6/8]...
Evaluating model pd [7/8]...
Evaluating model spd [8/8]...
Removing 2 genes with unresolved models
Comparative analysis of the 19559 without intronic signals.
Evaluating model no-reg [1/4]...
Evaluating model s [2/4]...
Evaluating model d [3/4]...
Evaluating model sd [4/4]...

The updated version (1.19.1) is already on GitHub and you can install it on your local computer by:

> library(devtools)
> install_github("ste-depo/INSPEcT")

Let me know if you meet additional problems with the code.

All the best,
Stefano

**Comment by slebedeva** [comment]
Thank you very much, Stefano!
Your implementation worked, but unfortunately more errors still exist..
You should be able to reproduce this with the above uploaded object:

```
diffrates12_48 <- compareSteady(SteadyObj[,c("12","48")])
Comparative analysis of the 72649 with both intronic and exonic signals.
Evaluating model no-reg [1/8]...
Evaluating model s [2/8]...
Evaluating model p [3/8]...
Evaluating model d [4/8]...
Evaluating model sp [5/8]...
Evaluating model sd [6/8]...
Evaluating model pd [7/8]...
Evaluating model spd [8/8]...
Removing 1 genes with unresolved models
Comparative analysis of the 19589 without intronic signals.
Evaluating model no-reg [1/4]...
Evaluating model s [2/4]...
Evaluating model d [3/4]...
Evaluating model sd [4/4]...
Removing 1 genes with unresolved models
Error in `.rowNamesDF<-`(x, value = value) : invalid 'row.names' length
```

Trying to subset gives:
```
compareSteady(SteadyObj[5391:5394,c("12","48")])
Comparative analysis of the 3 with both intronic and exonic signals.
Evaluating model no-reg [1/8]...
Evaluating model s [2/8]...
Evaluating model p [3/8]...
Evaluating model d [4/8]...
Evaluating model sp [5/8]...
Evaluating model sd [6/8]...
Evaluating model pd [7/8]...
Evaluating model spd [8/8]...
Comparative analysis of the 1 without intronic signals.
Evaluating model no-reg [1/4]...
Error in apply(outK_K[, 1:2] < 0, 1, any) : 
  dim(X) must have a positive length
In addition: Warning messages:
1: In min(wt_rates[wt_rates[, 1] != 0 & !is.na(wt_rates[, 1]), 1]) :
  no non-missing arguments to min; returning Inf
2: In min(wt_rates[wt_rates[, 2] != 0 & !is.na(wt_rates[, 2]), 2]) :
  no non-missing arguments to min; returning Inf
3: In max(wt_rates[wt_rates[, 1] != Inf, 1]) :
  no non-missing arguments to max; returning -Inf
4: In max(wt_rates[wt_rates[, 2] != Inf, 2]) :
  no non-missing arguments to max; returning -Inf
5: In min(sh_rates[sh_rates[, 1] != 0 & !is.na(sh_rates[, 1]), 1]) :
  no non-missing arguments to min; returning Inf
6: In min(sh_rates[sh_rates[, 2] != 0 & !is.na(sh_rates[, 2]), 2]) :
  no non-missing arguments to min; returning Inf
7: In max(sh_rates[sh_rates[, 1] != Inf, 1]) :
  no non-missing arguments to max; returning -Inf
8: In max(sh_rates[sh_rates[, 2] != Inf, 2]) :
  no non-missing arguments to max; returning -Inf
```

From the messages it seems that the problem might be with intronless genes?

Thanks a lot again for your time!
Svetlana

**Comment by ste-depo** [comment]
Hi Svetlana,

Those bugs are also fixed (version 1.19.1, commit 4831df1).

Let me know if you require more assistance!

Best,
Stefano

**Comment by slebedeva** [comment]
It works :-) Thank you very much Stefano!!!
